builtin/logical/jwt/path_issue.go

```diff
@@ -0,0 +1,173 @@
+				Description: "The time the JWT was issued",
+			},
+			"jti": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Default:     uuid.GenerateUUID(),
+				Description: "Unique identifier for the JWT",
+			},
+			"claims": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "JSON Object of Claims for the JWT Token",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.WriteOperation: b.pathIssueWrite,
+		},
+	}
+}
+
+func (b *backend) pathIssueWrite(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role").(string)
+	
+	// Get the role
+	role, err := b.getRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse(fmt.Sprintf("Unknown role: %s", roleName)), nil
+	}
+
+	token := jwt.New(jwt.SigningMethodRS256)
+
+	if data.Get("claims").(string) != "" {
+		// Parse JSON using unmarshal
+		var uc map[string]interface{}
+		err := json.Unmarshal([]byte(data.Get("claims").(string)), &uc)
+		if err != nil {
+			return nil, err
+		}
+		
+		for k, v := range uc {
+			token.Claims[k] = v
+		}
+	}
+
+	if _, ok := token.Claims["iss"]; !ok {
+		if data.Get("iss") != "" {
+			token.Claims["iss"] = data.Get("iss").(string)
+		}
+	}
+	if _, ok := token.Claims["sub"]; !ok {
+		if data.Get("sub") != "" {
+			token.Claims["sub"] = data.Get("sub").(string)
+		}
+	}
+	if _, ok := token.Claims["aud"]; !ok {
+		if data.Get("aud") != "" {
+			token.Claims["aud"] = data.Get("aud").(string)
+		}
+	}
+	if _, ok := token.Claims["exp"]; !ok {
+		if data.Get("exp").(int) > 0 {
+			token.Claims["exp"] = data.Get("exp").(int)
+		}
+	}
+	if _, ok := token.Claims["nbf"]; !ok {
+		if data.Get("nbf").(int) > 0 {
+			token.Claims["nbf"] = data.Get("nbf").(int)
+		}
+	}
+	if _, ok := token.Claims["iat"]; !ok {
+		if data.Get("iat").(int) > 0 {
+			token.Claims["iat"] = data.Get("iat").(int)
+		}
+	}
+	if _, ok := token.Claims["jti"]; !ok {
+		if data.Get("jti") != "" {
+			token.Claims["jti"] = data.Get("jti").(string)
+		}
+	}
+
+	entry, err := req.Storage.Get("config/" + roleName)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, fmt.Errorf(
+			"License Private Key haven't been configured. Please configure\n" +
+				"them at the 'config/" + roleName + "' endpoint")
+	}
+
+	var config roleConfig
+	if err := entry.DecodeJSON(&config); err != nil {
+		return nil, fmt.Errorf("error reading jwt configuration: %s", err)
+	}
+
+	tokenString, err := token.SignedString([]byte(config.Key))
+
+	if err != nil {
+		return nil, err
+	}
+
+	resp := b.Secret(SecretTokensType).Response(map[string]interface{}{
+		"jti": token.Claims["jti"].(string),
+		"token": tokenString,
+	}, map[string]interface{} {
+		"jti": data.Get("jti").(string),
+	})
+
+	err = req.Storage.Put(&logical.StorageEntry{
+		Key:   "tokens/" + data.Get("jti").(string),
```

Comment:
That's what I thought, please see my other comment on the main thread, advice on supporting renewals or not would be appreciated. 

> On Jul 30, 2015, at 10:41, Armon Dadgar <notifications@github.com> wrote:
> 
> In builtin/logical/jwt/path_issue.go:
> 
> > +
> > +	tokenString, err := token.SignedString([]byte(config.Key))
> > +
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +
> > +	resp := b.Secret(SecretTokensType).Response(map[string]interface{}{
> > +		"jti": token.Claims["jti"].(string),
> > +		"token": tokenString,
> > +	}, map[string]interface{} {
> > +		"jti": data.Get("jti").(string),
> > +	})
> > +
> > +	err = req.Storage.Put(&logical.StorageEntry{
> > +		Key:   "tokens/" + data.Get("jti").(string),
> You need to implement the Revoke on the Secret to delete this and respond with a lease value. If there is a lease, Vault will track the response and invoke Revoke on expiration. You could also support Renew so that the GC of the jti can be deferred.
> 
> —
> Reply to this email directly or view it on GitHub.
>